### PR TITLE
network: add P2PNetwork implementation

### PR DESCRIPTION
**Pull request by cce**
## Summary

Following up on #5634, this adds a new GossipNode implementation (P2PNetwork) which integrates with the libp2p support packages in network/p2p.

## Test Plan

Added new tests TestP2PSubmitTX and TestP2PSubmitWS, also tested nodes connecting with each other manually using `goal node start -p`. Added a basic 10-round test in test/e2e-go/features/p2p package along with 2-, 3-, and 5-node templates to start E2E testing.

### CODE REVIEW

File: network/p2pPeer.go
Comment (shiqizng) [inline network/p2pPeer.go:41]: the first 4 bytes stores the length? can we add a reference here please?

File: cmd/algod/main.go
Comment (shiqizng) [inline cmd/algod/main.go:289]: peerOverrideArray could include both multiaddr and  url.Host. will nodes be able to resolve both of these addresses?

File: network/p2p/pubsub.go
Comment (Eric-Warehime) [inline network/p2p/pubsub.go:82]: It feels like we should move topic-specific params/functions into their own module. I can envision something like `network/p2p/topics/tx.go`, `network/p2p/topics/agreement.go`, etc.

**Review by Eric-Warehime — COMMENTED** [review]
Nice start, I think this will get us most of the way to an initial test!

Some general comments at this point.
* I think we should remove all of the Todo comments. Let's capture them in issues if we need to, and make sure things are relatively clean as we commit them.
* Needs more test coverage.
* Might be helpful to commit alongside this some readme or diagram which explains how things are structured in terms of the different gossipsub topics/what they are used for, and the things we're running through streams.

**Comment by pao-beep** [comment]
I support a diagram, I hope it's public

File: network/p2p/pubsub.go
Comment (iansuvak) [reply network/p2p/pubsub.go:50]: I think it's just a starting place to be as permissive as possible for the initial tests

File: network/p2p/p2p.go
Comment (iansuvak) [reply network/p2p/p2p.go:121]: It might be possible to have connections exceed the target given multiple ways of connecting to others and the possibility of others connecting to us. I think it would be better to have this more robust condition than strict equality.

File: network/p2p/peerstore/peerstore.go
Comment (shiqizng) [reply network/p2p/peerstore/peerstore.go:31]: ok. that's too bad. I wanted to avoid writing those wrapper methods.

File: network/p2p/pubsub.go
Comment (iansuvak) [reply network/p2p/pubsub.go:50]: This sounds reasonable to me. We are likely going to tweak this in the future but this proposal sounds as good as any for now.

File: cmd/algod/main.go
Comment (iansuvak) [reply cmd/algod/main.go:289]: I think logging an error is preferable to bringing the node down for a single typo in a single manual entry. 

I have a PR against this branch that potentially addresses this: https://github.com/cce/go-algorand/pull/15 but I guess it's not useful at all if we implement a third GossipNode for the hybrid version

File: network/wsNetwork.go
Comment (Eric-Warehime) [reply network/wsNetwork.go]: It doesn't have to be a separate package, I just find as a maintainer tracking behavior in files multiple thousand lines long begins to slow me down when editing existing code.

File: tools/block-generator/go.mod
Comment (Eric-Warehime) [reply tools/block-generator/go.mod:27]: Got it. Thanks for explaining!

File: config/local_defaults.go
Comment (Eric-Warehime) [reply config/local_defaults.go:72]: Doesn't matter to me either way in the end. I think the convention makes it very hard to determine  which properties are related or interact with one another as some properties do. 

We would probably need to do a bigger refactor of the config system to solve problems like this in the future though.

**Review by Eric-Warehime — APPROVED** [review]
Unit test coverage is still quite low, but I guess it's ok as a starting place.

File: network/p2pNetwork.go
Comment (shiqizng) [inline network/p2pNetwork.go:281]: will this be implemented eventually to support catchpoint catchup?

Comment (iansuvak) [reply network/p2pNetwork.go:281]: Potentially yeah, right now it's just left trivial to implement the GossipNode interface.

File: network/p2pNetwork.go
Comment (algorandskiy) [inline network/p2pNetwork.go:234]: this mean broadcast requests to p2pnet both published to gossib sub and pushed to neighbors?

File: network/p2p/streams.go
Comment (iansuvak) [reply network/p2p/streams.go:94]: peer.ID is just a string so this is regular string ordering but peer.ID is validated at handshake time

File: network/p2p/pubsub.go
Comment (iansuvak) [reply network/p2p/pubsub.go]: Renamed the method and made it readlocking for most cases.

File: network/p2pNetwork.go
Comment (iansuvak) [reply network/p2pNetwork.go:44]: Done, although now wsNetwork uses a constant defined in p2pNetwork.go file (still in the network package though). I still think we should create a general network file defining the GossipNode interface and relevant constants used by both wsNetwork and p2pNetwork implementation but that can wait post merge of this PR.

File: netdeploy/remote/nodeConfig.go
Comment (iansuvak) [reply netdeploy/remote/nodeConfig.go]: Made the change to use string in this pacakge since peer.ID is a type alias for string and removed it's use in the netdeploy package as well by creating a p2p package local notion of peerID.

File: network/p2p/streams.go
Comment (algorandskiy) [reply network/p2p/streams.go:94]: I mean I do not understand why/how to compare local vs remote peer IDs

**Review by algorandskiy — DISMISSED** [review]
Please re-merge master and fix "const `peerStorePath` is unused (unused)" linter.

**Comment by iansuvak** [comment]
Remerged master and disabled the FiveNode E2E test. It's currently failing because of a timing issue that nodes are unable to recover from because HTTP block service is not yet implemented. This will be solved in a future PR.

File: test/e2e-go/features/p2p/p2p_basic_test.go
Comment (algorandskiy) [inline test/e2e-go/features/p2p/p2p_basic_test.go:66]: Add a s
```suggestion
	t.Skip("Disabled due to block service timeouts. TODO: fix/reenable")
```

**Review by algorandskiy — APPROVED** [review]

**Review by shiqizng — APPROVED** [review]

**Review by iansuvak — APPROVED** [review]
Approving as a second reviewer since @algorandskiy approved recently and the only change since was the readme and a single line suggestion by him.

File: network/p2pNetwork.go
Comment (iansuvak) [inline network/p2pNetwork.go:261]: Spoke with @cce offline but logging here anyway. This looks like it should work but empirically only adding peers for PeersConnectedOut and PeersConnectedIn doesn't fix the test but adding it for all peer options does. 

Tracking it down now but we should hold off on merging until we figure it out.

**Review by iansuvak — APPROVED** [review]

**Review by shiqizng — DISMISSED** [review]

File: test/e2e-go/features/p2p/p2p_basic_test.go
Comment (algorandskiy) [inline test/e2e-go/features/p2p/p2p_basic_test.go:66]: why removed the comment?

**Review by algorandskiy — APPROVED** [review]

**Review by iansuvak — APPROVED** [review]
LGTM. I like the move of general code gossipNode.go

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**
